Запрос рекомендации

ahim-miha opened the pull request:
https://faang-school.atlassian.net/browse/BJS2-40432 
1.
create(RecommendationRequestDto recommendationRequestDto): Создает новый запрос на рекомендацию.
Валидирует входные данные с помощью RecommendationRequestValidator. Преобразует DTO в сущность и сохраняет её в репозитории. Сохраняет связанные навыки через SkillRequestService. 
2.
getRequest(RecommendationRequestFilterDto recommendationRequestFilterDto): Извлекает запросы на рекомендации, применяя фильтры. Валидирует доступность фильтров.
Применяет фильтры к списку запросов.

3.
getRequest(long id):
Извлекает запрос на рекомендацию по идентификатору. Если запрос не найден, выбрасывает исключение DataValidationException. 
4.
rejectRequest(long id, RejectionDto rejectionDto): Отклоняет запрос на рекомендацию по идентификатору. Валидирует существование запроса.
Устанавливает статус REJECTED и причину отклонения. Сохраняет изменения в репозитории.

ahim-miha committed:
Запрос рекомендации
https://faang-school.atlassian.net/browse/BJS2-40432
1.
create(RecommendationRequestDto recommendationRequestDto):
Создает новый запрос на рекомендацию.
Валидирует входные данные с помощью RecommendationRequestValidator.
Преобразует DTO в сущность и сохраняет её в репозитории.
Сохраняет связанные навыки через SkillRequestService.
Возвращает DTO сохраненного запроса.
2.
getRequest(RecommendationRequestFilterDto recommendationRequestFilterDto):
Извлекает запросы на рекомендации, применяя фильтры.
Валидирует доступность фильтров.
Применяет фильтры к списку запросов.
Возвращает список DTO отфильтрованных запросов.
3.
getRequest(long id):
Извлекает запрос на рекомендацию по идентификатору.
Если запрос не найден, выбрасывает исключение DataValidationException.
Возвращает DTO найденного запроса.
4.
rejectRequest(long id, RejectionDto rejectionDto):
Отклоняет запрос на рекомендацию по идентификатору.
Валидирует существование запроса.
Устанавливает статус REJECTED и причину отклонения.
Сохраняет изменения в репозитории.
Возвращает DTO отклоненного запроса.

Memoreks commented on src/main/java/school/faang/user_service/controller/recommendationRequest/RecommendationRequestController.java:
```diff
@@ -0,0 +1,40 @@
+@RequiredArgsConstructor
+public class RecommendationRequestController {
+
+    public final RecommendationRequestService recommendationRequestService;
```
public?

Memoreks commented on src/main/java/school/faang/user_service/dto/recommendationRequest/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,23 @@
+    private Long id;
+    private String message;
+    private RequestStatus status;
+    private List<Long> skillList;
```
Может быть лучше назвать skills? Тогда и в маппере не надо будет через Mapping сопоставлять

Memoreks commented on src/main/java/school/faang/user_service/validator/recommendationRequest/RecommendationRequestValidator.java:
```diff
@@ -0,0 +1,105 @@
+public class RecommendationRequestValidator {
+    private static final int SIX_MONTHS_AGO = 6;
+
+    public void validateCreate(RecommendationRequestDto recommendationRequestDto,
+                               RecommendationRequestRepository recommendationRequestRepository,
+                               UserRepository userRepository,
+                               SkillRepository skillRepository) {
```
Может быть лучше создать внутри RecommendationRequestValidator все репозитории, чтобы каждый раз их не передавать в метод?

Memoreks commented on src/test/java/school/faang/user_service/service/controller/SkillRequestService/SkillRequestServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+    @Mock
+    private SkillRepository skillRepository;
+
+    @BeforeEach
+    void setUp() {
+
+    }
```
А зачем пустой метод? Туда бы добавить объекты, которые используются в тестах

Memoreks commented on src/test/java/school/faang/user_service/service/controller/recommendationRequest/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,217 @@
+        requester = User.builder()
+                .id(1L)
+                .build();
+        receiver = User.builder()
```
Я бы добавил отступ, чтобы код не сливался

Memoreks commented on src/test/java/school/faang/user_service/service/controller/recommendationRequest/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,217 @@
+    @DisplayName("Receive recommendation requests with filters")
+    public void requestRecommendationsWithFiltersTest() {
+        List<RecommendationRequest> recommendationRequestsList = new ArrayList<>();
+        recommendationRequestsList.add(RecommendationRequest.builder()
+                .id(1L)
+                .requester(requester)
+                .receiver(receiver)
+                .skills(List.of(new SkillRequest(), new SkillRequest()))
+                .build()
+        );
+        recommendationRequestsList.add(RecommendationRequest.builder()
+                .id(1L)
+                .requester(requester)
+                .receiver(receiver)
+                .skills(List.of(new SkillRequest(), new SkillRequest()))
+                .build()
+        );
```
Лучше бы избавиться от дублирования кода

Memoreks commented on src/test/java/school/faang/user_service/service/controller/validator/recommendationRequest/RecommendationRequestValidatorTest.java:
```diff
@@ -0,0 +1,195 @@
+                .skillList(List.of(1L, 2L, 3L))
+                .createdAt(LocalDateTime.now())
+                .build();
+
+
```
Лишние отступы

Memoreks commented on src/test/java/school/faang/user_service/service/controller/validator/recommendationRequest/RecommendationRequestValidatorTest.java:
```diff
@@ -0,0 +1,195 @@
+    @DisplayName("Validate no recent request")
+    public void validateNoRecentRequestTest() {
+        RecommendationRequest request = new RecommendationRequest();
+        request.setCreatedAt(LocalDateTime.now().minusMonths(7));
```
Я бы вынес все "магические" цифры в коде в отдельные переменные, для улучшения читабельности и понимания кода

Memoreks reviewed (commented):
Хорошо написал код. По нему удобно ориентироваться и многие вещи понятны. Оставил несколько комментариев, которые повысят читабельность и понимание кода.

VovaBuzivskyi commented on src/main/java/school/faang/user_service/dto/recommendationRequest/RecommendationRequestFilterDto.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.time.LocalDateTime;
+
+@Data
```
Нужно ли здесь ставить Data? Я так понимаю здесь тебе нужны геттеры сеттеры и ToString.

VovaBuzivskyi commented on src/main/java/school/faang/user_service/dto/recommendationRequest/RejectionDto.java:
```diff
@@ -0,0 +1,10 @@
+import lombok.Builder;
+import lombok.Data;
+
+@Data
```
Здесь тоже самое, только сетеры и гетеры нужны.

VovaBuzivskyi commented on src/main/java/school/faang/user_service/service/SkillRequestService/SkillRequestService.java:
```diff
@@ -0,0 +1,29 @@
+    public void saveSkillRequests(RecommendationRequest recommendationRequestEntity, List<Long> skillIds) {
+        List<SkillRequest> skillRequests = skillIds.stream()
+                .map(skillId -> new SkillRequest(0, recommendationRequestEntity, skillRepository.findById(skillId)
+                        .orElseThrow()))
```
Может здесь стоит выбросить сообщение о том что скилл с таким то Id не был найден.

VovaBuzivskyi commented on src/test/java/school/faang/user_service/service/controller/validator/recommendationRequest/RecommendationRequestValidatorTest.java:
```diff
@@ -0,0 +1,195 @@
+
+    @Test
+    @DisplayName("Validate skills with valid data")
+    public void vValidateSkillsWithValidDataTest() {
```
Лишняя буква.

VovaBuzivskyi commented:
Код написан хорошо. Оставил мелкие замечания.

ahim-miha committed:
Запрос рекомендации
https://faang-school.atlassian.net/browse/BJS2-40432
Исправил ошибки

BobrAll commented on src/main/java/school/faang/user_service/dto/recommendationRequest/RejectionDto.java:
```diff
@@ -0,0 +1,14 @@
+@Setter
+@ToString
+@Builder
+public class RejectionDto {
```
Лучше бы RecommendationRejectionDto тогда

BobrAll commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
Круто

BobrAll commented on src/main/java/school/faang/user_service/validator/recommendationRequest/RecommendationRequestValidator.java:
```diff
@@ -0,0 +1,109 @@
+@RequiredArgsConstructor
+public class RecommendationRequestValidator {
+
+    private static final int SIX_MONTHS_AGO = 6;
```
Хорошо, что в константу вынес, но название не совсем понятное, если увидеть его вне контекста. Как насчет MONTHS_TO_REPEAT_REQUEST?

BobrAll commented on src/test/java/school/faang/user_service/service/controller/SkillRequestService/SkillRequestServiceTest.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Test
+    @DisplayName("Check For Lack Of Skills")
+    public void checkForLackOfSkillsTest() {
```
Можно еще проверить, что у репозитория не вызывается метод save, чтобы точно убедиться, что мы перед вылетом исключения не сохранили ничего лишнего

BobrAll reviewed (changes requested):
Наверное этот модуль для тебя намного интереснее, сильно выросло качество кода, прям приятно смотреть! Оставил пару правок - они совсем незначительные. В любом есть конфликт, который нужно решить, заодно лучше и их поправить

BobrAll commented:
И отдельный лайк за @Valid

ahim-miha committed:
Запрос рекомендации
https://faang-school.atlassian.net/browse/BJS2-40432
Исправил ошибки

BobrAll commented:
Все супер, но merge-конфликт остался, он не позволяет влить изменения

ahim-miha committed:
Merge branch 'Manticore-master-stream7' into feature-BJS2-40432

BobrAll commented:
Тесты не проходят. Либо стоит подтянуть изменения актуальные, либо какие-то ошибки в коде

<img width="921" alt="image" src="https://github.com/user-attachments/assets/b53ae950-54e3-4179-9ebf-93c797dcbb83">

ahim-miha committed:
Запрос рекомендации
https://faang-school.atlassian.net/browse/BJS2-40432
Починил все тесты, переписал метод с фильтром и немного переделал фильтры

ahim-miha committed:
Запрос рекомендации
https://faang-school.atlassian.net/browse/BJS2-40432
Починил все тесты, переписал метод с фильтром и немного переделал фильтры

ahim-miha committed:
Merge remote-tracking branch 'origin/feature-BJS2-40432' into feature-BJS2-40432

BobrAll reviewed (commented):
Отлично, вливаю
